AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/InitialSync/ModelSyncedEventEmitter.swift

```diff
@@ -0,0 +1,118 @@
+//
+// Copyright 2018-2020 Amazon.com,
+// Inc. or its affiliates. All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Amplify
+import AWSPluginsCore
+import Combine
+
+@available(iOS 13.0, *)
+final class ModelSyncedEventEmitter {
+    var syncOrchestratorSink: AnyCancellable?
+    var reconciliationQueueSink: AnyCancellable?
+
+    let modelType: Model.Type
+    var recordsReceived: AtomicValue<Int>
+    var reconciledReceived: AtomicValue<Int>
+    var initialSyncOperationFinished: Bool
+
+    private var modelSyncedEventBuilder: ModelSyncedEvent.Builder
+
+    private var modelSyncedEventTopic: PassthroughSubject<Never, Never>
+    var publisher: AnyPublisher<Never, Never> {
+        return modelSyncedEventTopic.eraseToAnyPublisher()
+    }
+
+    init(modelType: Model.Type,
+         initialSyncOrchestrator: InitialSyncOrchestrator?,
+         reconciliationQueue: IncomingEventReconciliationQueue?) {
+        self.modelType = modelType
+        self.recordsReceived = AtomicValue(initialValue: 0)
+        self.reconciledReceived = AtomicValue(initialValue: 0)
+        self.initialSyncOperationFinished = false
+
+        self.modelSyncedEventBuilder = ModelSyncedEvent.Builder()
+
+        self.modelSyncedEventTopic = PassthroughSubject<Never, Never>()
+
+        self.syncOrchestratorSink = initialSyncOrchestrator?
+            .publisher
+            .sink(receiveCompletion: { _ in },
+                  receiveValue: { [weak self] value in
+                    self?.onReceiveSyncOperationEvent(value: value)
+            })
+
+        self.reconciliationQueueSink = reconciliationQueue?
+            .publisher
+            .sink(receiveCompletion: { _ in },
+                  receiveValue: { [weak self] value in
+                    self?.onReceiveReconciliationEvent(value: value)
+            })
+    }
+
+    private func onReceiveSyncOperationEvent(value: InitialSyncOperationEvent) {
+        switch value {
+        case .started(let modelType, let syncType):
+            guard self.modelType == modelType else {
+                return
+            }
+            modelSyncedEventBuilder.isFullSync = syncType == .fullSync ? true : false
+            modelSyncedEventBuilder.isDeltaSync = !modelSyncedEventBuilder.isFullSync
```

Comment:
As discussed, I agree that these will only be mutated when receiving a `.started` event, which should only happen once. If it were me, I'd probably still make the change to have one queue owned by ModelSyncedEventEmitter and make sure all my accesses happen on it, rather than scattering `AtomicValues` throughout the code, which proliferates queues and makes the code more brittle for future changes. But it's not something I'll block you on now, as long as you're able to reason that this is safe, and prove it with tests.

Reply:
I updated `ModelSyncedEventEmitter` to use `DispatchQueue`, please take a look if I am using it correctly. Thanks